fix(lsp): robust rename()

tomtomjhj opened the pull request:
* Match the prefix of the list of the path components.
* Directly rename the existing buffers without deleting them, so that states like undo, window layout, etc are preserved.
* Don't write if not writable.
* buffer:// was removed in 236c20795eb9f11e21e0719b735ea741711acc08.
* Remove redundant custom bufwinid function.
* Fix tests.

glepnir commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -771,72 +771,97 @@ function M.text_document_completion_list_to_complete_items(result, prefix)
+  return result
+end
+
+local function escape_gsub_repl(s)
```
does there can't use `vim.fn.escape` ?

tomtomjhj replied:
lua-pattern and gsub {repl} are escaped with `%`, but escape() uses `\`

zeertzjq replied:
Does `vim.psub()` not work?

glepnir replied:
forget `vim.pesc` then.

zeertzjq commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -771,72 +771,97 @@ function M.text_document_completion_list_to_complete_items(result, prefix)
   return buffers
 end
 
+local function escape_gsub_pat(s)
```
`vim.pesc()`

mfussenegger reviewed (commented):
These changes make the behavior actually worse for me. If I rename a class with eclipse.jdt.ls I am afterwards left with error diagnostics because of missing document change notifications.

Could you explain a bit more what problem you're trying to solve here?

tomtomjhj commented:
The old implementation deletes the buffers that are affected by renaming. Then you will lose the undo information for that buffer, and the windows displaying that buffer will be closed (among many other side effects). I'd like to retain these states even after renaming. 

If you think this is useful, I'll try to fix the missing notification problem. Otherwise, I'll just remove the "Directly rename the existing buffers without deleting them" part of the change.

tomtomjhj commented:
Oh there's another problem with this change. After renaming, `:write`-ing in the renamed buffer raises E13.

tomtomjhj commented:
Fixed E13 issue. 

> If I rename a class with eclipse.jdt.ls I am afterwards left with error diagnostics because of missing document change notifications.

@mfussenegger Do you disable didChangeWatchedFiles (as discussed in #23291)? With this enabled, at least rust-analyzer seems to pick up the changes automatically.

mfussenegger commented:
>  Do you disable didChangeWatchedFiles

No I have it enabled.

Based on the specification I'd think it's important that the `didClose` / `didOpen` and other document synchronization messages happen. If this PR prevents that I don't think we can merge it.

tomtomjhj commented:
I have modified the implementation to utilize the existing `didClose`/`didOpen` logic for `:saveas`.

przepompownia commented:
@tomtomjhj could you to merge the current master here? 

Currently on this branch after calling `vim.lsp.util.rename('src/Old.php', 'src/New.php')` both files exists and `Old.php` remains the current buffer in the window (like before call). 

I ended up here by #27291 but am also interested in what it does in practice. I try to debug why the index of phpactor LSP is outdated after renaming a class. If I make this call on master nvim, I see four client notifications:

```log
"rpc.send" { jsonrpc = "2.0", method = "textDocument/didSave", params = { textDocument = { uri = "file:///.../src/Old.php" } }}
"rpc.send" { jsonrpc = "2.0", method = "textDocument/didOpen", params = { textDocument = { languageId = "php", text = "<?php\n\nnamespace ... \n", uri = "file:///.../src/New.php", version = 0 } }}
"rpc.send" { jsonrpc = "2.0", method = "textDocument/didClose", params = { textDocument = { uri = "file:///.../src/Old.php" } }}
"rpc.send" { jsonrpc = "2.0", method = "workspace/didChangeWatchedFiles", params = { changes = { { type = 3, uri = "file:///.../src/Old.php" }, { type = 1, uri = "file:///.../src/New.php" } } }}

```
Is there `textDocument/didSave` for the new file missing?

tomtomjhj commented:
rebased

> Currently on this branch after calling `vim.lsp.util.rename('src/Old.php', 'src/New.php')` both files exists and `Old.php` remains the current buffer in the window (like before call).

the issue seems to be that it works only with normalized absolute path

tomtomjhj commented:
> the issue seems to be that it works only with normalized absolute path

fixed

przepompownia commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -639,76 +639,108 @@ function M.text_document_completion_list_to_complete_items(result, prefix)
+
+  local old_fname_full = vim.uv.fs_realpath(vim.fs.normalize(old_fname, { expand_env = false }))
+  if not old_fname_full then
+    vim.notify('Invalid path: ' .. old_fname)
```
```suggestion
    vim.notify('Invalid path: ' .. old_fname, vim.log.levels.ERROR)
```

przepompownia commented:
With the recent change I get 
```log
"rpc.send" { jsonrpc = "2.0", method = "textDocument/didClose", params = { textDocument = { uri = "file:///.../src/Old.php" } }}
"rpc.send" { jsonrpc = "2.0", method = "textDocument/didOpen", params = { textDocument = { languageId = "php", text = "<?php\n\nnamespace ...\n", uri = "file:///.../src/New.php", version = 0 } }}
"rpc.send" { jsonrpc = "2.0", method = "textDocument/didSave", params = { textDocument = { uri = "file:///.../src/New.php" } }}
"rpc.send" { jsonrpc = "2.0", method = "workspace/didChangeWatchedFiles", params = { changes = { { type = 3, uri = "file:///.../src/Old.php" }, { type = 2, uri = "file:///.../src/New.php" } } }}
```
I didn't check yet if it (especially `didSave` notification for the new file) can fix non-trivial rename operation in phpactor.

przepompownia commented:
Probably using `:saveas` here solves the same problem as in #27408: the new buffer becomes visible on buffer list (`:ls` without `!`) and I can alternate it with another buffer. Consider tests it from like in #27408 or another way.

mfussenegger reviewed (commented):
I'm sorry but I still don't really understand the problem(s?) that this PR solves.

I'd rather go with more targeted and smaller PRs like https://github.com/neovim/neovim/pull/27408 instead which I can reason about more easily.

tomtomjhj committed:
refactor(lsp): remove redundant code

* use builtin function
* buffer:// was removed in 236c20795eb9f11e21e0719b735ea741711acc08.

tomtomjhj committed:
fix(lsp): normalize and check rename target

tomtomjhj committed:
fix(lsp): when renaming directory, check path prefix of buffer names

For example, when renaming /path/to/dir, buffers like
fern://drawer/file:///path/to/dir, /path/to/dir123 should not be
matched.

tomtomjhj committed:
perf(lsp): remove redundant buf load/write

tomtomjhj committed:
fix(lsp): make sure that renaming doesn't overwrite buffer without file

tomtomjhj committed:
fix(lsp): directly rename the existing buffers

Problem:
Renaming a file deletes the buffers that are affected by renaming. This
has undesireable side effects. For example, the undo information is lost
for those buffers, and the windows displaying those buffers are closed.

Solution:
Rename the buffers with :saveas or :file_f.

tomtomjhj committed:
test(lsp): test rename()

Also remove tests that doesn't conform to the new semantics:
* `rename()` maintains the identity of the renamed buffers.
* `rename(old, new)` when there are buffer for both `old` and `new`:
  Normally, such a rename is not allowed.
  But If `new` is unloaded, it's safe to delete `new`.

przepompownia commented:
Nice work, worth testing some of the cases described in individual commits. 

> I have split the commits so that each of them fixes one problem.

tomtomjhj committed:
docs(lsp): clarify rename() behavior

tomtomjhj commented:
I think now the test covers most of the important cases in the `fix` commits.

tomtomjhj commented:
I'll make prs for smaller units of changes
